### PR TITLE
Read Multiple Values for Common Key into Array

### PR DESCRIPTION
**Pull request by heilkn**
Dear lipkau,

thank you very much for the great tool. When I wanted to use it, I encountered the problem that the ini files I have to modify contain several keys in a section with the same name. Therefore, I'd like to propose to read in those entries as an array.

Please consider adding this feature. This pullrequest is a simple approach to solve that problem. Probably, Set-IniContent and Out-IniFile needs adaption, too. I think you will know best which adjustments need to be made. However, if you are not able to do the necessary changes, please let me know what is required.

Kind regards

Konstantin

### CODE REVIEW

File: PSIni/Functions/Out-IniFile.ps1 [hunk 1]
```diff
@@ -194,13 +200,13 @@ Function Out-IniFile {
...
             }
-            elseif ($i -eq $script:NoSection) {
-                #Key value pair of NoSection
-                Out-Keys $InputObject[$i] `
-                    @parameters `
-                    -Delimiter $delimiter `
-                    -MyInvocation $MyInvocation
-            }
+#            elseif ($i -eq $script:NoSection) {
+#                #Key value pair of NoSection
```
Comment (lipkau) [inline PSIni/Functions/Out-IniFile.ps1]: Why are you commenting this part out?

File: PSIni/Functions/Get-IniContent.ps1 [hunk 2]
```diff
@@ -141,7 +141,15 @@ Function Get-IniContent {
                 }
                 $name, $value = $matches[1..2]
                 Write-Verbose "$($MyInvocation.MyCommand.Name):: Adding key $name with value: $value"
-                $ini[$section][$name] = $value
+                if ($ini[$section][$name] -is [string]) {
+                    $ini[$section][$name] = $ini[$section][$name], $value
```
Comment (lipkau) [inline PSIni/Functions/Get-IniContent.ps1]: Why are you appending as array, if it's a string?

File: PSIni/Functions/Out-IniFile.ps1 (hunk 1 above)
Comment (heilkn) [reply PSIni/Functions/Out-IniFile.ps1]: Uh, I am deeply sorry, about this. This is a remained of a debugging step.

File: PSIni/Functions/Get-IniContent.ps1 (hunk 2 above)
Comment (heilkn) [reply PSIni/Functions/Get-IniContent.ps1]: This is the strategy. If the content is only one value, then it is a simple string. If a second value needs to be added, the entry is converted to an array. If you have a better idea how to handle this, please let me know.

**Comment by heilkn** [comment]
Regarding the duplicate names: I think it is up to the program how to interprete the ini file. However, when dealing with ini files on the command line, you usually want to do changes to an existing file. If you know, a program does not support multiple entries, don't create any. If a program does, then the tool is of no help if it can't support that.

File: PSIni/Functions/Get-IniContent.ps1 (hunk 2 above)
Comment (lipkau) [reply PSIni/Functions/Get-IniContent.ps1]: after tinkering a bit with this, I would like you to replace the code with this:

```powershell
   if ( -not $ini[$section][$name]) {
        $ini[$section][$name] = @($value)
    }
    else {
        $ini[$section][$name] += $value
    }
```

File: PSIni/Functions/Out-IniFile.ps1 [hunk 3]
```diff
@@ -153,7 +153,13 @@ Function Out-IniFile {
                     }
                     else {
                         Write-Verbose "$($MyInvocation.MyCommand.Name):: Writing key: $key"
-                        Add-Content -Value "$key$delimiter$($InputObject[$key])" -Encoding $Encoding -Path $Path
+                        if ($InputObject[$key] -is [string]) {
```
Comment (lipkau) [inline PSIni/Functions/Out-IniFile.ps1]: After tinkering a bit with this, I would like you to replace the code with this:

```powershell
$InputObject[$key] |
        Foreach-Object { "$key$delimiter$($InputObject[$key])" } |
        Add-Content -Value  -Encoding $Encoding -Path $Path
```

**Comment by lipkau** [comment]
Also:
can you please add tests for this new behavior?

File: PSIni/Functions/Out-IniFile.ps1 (hunk 3 above)
Comment (heilkn) [reply PSIni/Functions/Out-IniFile.ps1]: done

File: PSIni/Functions/Get-IniContent.ps1 (hunk 2 above)
Comment (heilkn) [reply PSIni/Functions/Get-IniContent.ps1]: I have done as you suggested.

**Comment by heilkn** [comment]
@lipkau Please let me know if you think more work is required on this, before it can be merged.

File: Tests/PsIni.Tests.ps1 [hunk 4]
```diff
@@ -40,7 +40,7 @@ Describe "PsIni functionality" {
     $dictIn["Category1"]["Key1"] = "Value1"
     $dictIn["Category1"]["Key2"] = "Value2"
     $dictIn["Category2"] = New-Object System.Collections.Specialized.OrderedDictionary([System.StringComparer]::OrdinalIgnoreCase)
-    $dictIn["Category2"]["Key3"] = "Value3"
+    $dictIn["Category2"]["Key3"] = @("Value3.1", "Value3.2")
```
Comment (lipkau) [inline Tests/PsIni.Tests.ps1]: One last thing:
please make the sample with 3 entries
this makes clear, that it's a flat array (not `@(@("Value3.1"), "Value3.2")`)`

Comment (heilkn) [reply Tests/PsIni.Tests.ps1]: I have added a third value to "key3".

Thank you very much for your support!

**Comment by EliaSaSe** [comment]
Dear @lipkau 

FYI - Similar problem as with versions 1.2.38 / 42 and 2.0.0.0:

This change introduces a breaking change between version 2.0.5 and 2.0.6.
For sections with non-duplicate keys, there will be also returned an array instead of a string. 

**Example:**

*my.ini*
```
[Trace]
Path=C:\temp
PurgeAfterDays=5
Level=Warning
```

*behavior 2.0.5*
```powershell
> Import-Module PsIni -MaximumVersion 2.0.5
> $c = Get-IniContent -FilePath C:\temp\my.ini
> $c.Trace
Name                           Value
----                           -----
Path                           C:\temp
PurgeAfterDays                 5
Level                          Warning

> $c.Trace.Path.GetType()
IsPublic IsSerial Name   BaseType
-------- -------- ----   --------
True     True     String System.Object
```

*behavior 2.0.6*
```powershell
> Import-Module PsIni -MinimumVersion 2.0.6
> $c = Get-IniContent -FilePath C:\temp\my.ini
> $c.Trace

Name                           Value
----                           -----
Path                           {C:\temp}
PurgeAfterDays                 {5}
Level                          {Warning}

PS C:\Users\ese> $c.Trace.Path.GetType()

IsPublic IsSerial Name     BaseType
-------- -------- ----     --------
True     True     Object[] System.Array
```

Because I have to support PsIni version 2.x, I have to deal with both behaviors.
As workaround I need to test the value with *-is [array]*. But that's manageable.

Best regards
Elia

**Comment by EliaSaSe** [comment]
I expressed myself wrong: I don't have to support Powershell 2.0, I need to support **PsIni in Version 2.X**. I told my users: "To use my module, you need PsIni in Major/Minor-Version 2.0, because my module reads and writes ini-files."

I don't have control over all systems where my module is running. Some of my ini-related module-functions are now failling when someone updates PsIni from <=2.0.5 to 2.0.6 because of the behavior changes in this pull request.

In my next module version, I will use PsIni as nested module to be sure, which version of PsIni my module is using. This will remove the external dependency to PsIni.

**Comment by lipkau** [comment]
should be fixed
https://github.com/lipkau/PsIni/commit/c7b6b089dbd241d7fa8a9489a4b6d57db6c71664

**Comment by EliaSaSe** [comment]
Cool. Thank you.

**Pull request merged**